File: lua/tst/video.lua
require 'pico.check'

pico.init(true)
pico.set.window { title="Video" }
pico.set.color.clear({'!', r=0x00, g=0x00, b=0x00})

-- pico_get_video: check Y4M header parsing
do
    print("get_video: dimensions and fps")
    local info = pico.get.video("video.y4m")
    assert(info.dim.w == 10)
    assert(info.dim.h == 10)
    assert(info.fps == 10)
    assert(info.frame == 0)
    assert(info.done == false)
end

-- pico_layer_video: create layer
do
    print("layer_video: create")
    local name = pico.layer.video("vid", "video.y4m")
    assert(name == "vid")
end

-- pico_video_sync: first frame (0)
do
    print("video_sync: frame 0 (top-left)")
    pico.set.layer("vid")
    local ok = pico.video.sync("vid", 0)
    assert(ok == true)
    pico.set.layer(nil)
    pico.output.clear()
    pico.output.draw.layer("vid",
        {'%', x=0.5, y=0.5, w=1, h=1, anchor='C'})
    pico.check("video-01")
end

-- pico_video_sync: middle frame (25)
do
    print("video_sync: frame 25 (bottom-middle)")
    local ok = pico.video.sync("vid", 25)
    assert(ok == true)
    pico.output.clear()
    pico.output.draw.layer("vid",
        {'%', x=0.5, y=0.5, w=1, h=1, anchor='C'})
    pico.check("video-02")
end

-- pico_video_sync: last frame (49)
do
    print("video_sync: frame 49 (top-right)")
    local ok = pico.video.sync("vid", 49)
    assert(ok == true)
    pico.output.clear()
    pico.output.draw.layer("vid",
        {'%', x=0.5, y=0.5, w=1, h=1, anchor='C'})
    pico.check("video-03")
end

-- pico_video_sync: backward seek
do
    print("video_sync: backward seek to frame 10")
    local ok = pico.video.sync("vid", 10)
    assert(ok == true)
    pico.output.clear()
    pico.output.draw.layer("vid",
        {'%', x=0.5, y=0.5, w=1, h=1, anchor='C'})
    pico.check("video-04")
end

-- pico_video_sync: past EOF
do
    print("video_sync: past EOF returns 0")
    local ok = pico.video.sync("vid", 100)
    assert(ok == false)
end

-- pico_get_video: check state after sync
do
    print("get_video: state after sync")
    local info = pico.get.video("vid")
    assert(info.done == true)
end

-- pico_output_draw_video: all-in-one
do
    print("draw_video: first call")
    pico.output.clear()
    local ok = pico.output.draw.video("video.y4m",
        {'%', x=0.5, y=0.5, w=1, h=1, anchor='C'})
    assert(ok == true)
    pico.check("video-05")
end

pico.init(false)

File: tst/video.c
#include "pico.h"
#include "../check.h"

int main (void) {
    pico_init(1);
    pico_set_window("Video", -1, NULL);
    pico_set_color_clear((Pico_Color){0x00, 0x00, 0x00});

    /* pico_get_video: check Y4M header parsing */
    {
        puts("get_video: dimensions and fps");
        Pico_Video info = pico_get_video("video.y4m", NULL);
        assert(info.dim.w == 10);
        assert(info.dim.h == 10);
        assert(info.fps == 10);
        assert(info.frame == 0);
        assert(info.done == 0);
    }

    /* pico_layer_video: create layer */
    {
        puts("layer_video: create");
        const char* name = pico_layer_video("vid", "video.y4m");
        assert(strcmp(name, "vid") == 0);
    }

    /* pico_set_video: first frame (0) */
    {
        puts("set_video: frame 0 (top-left)");
        pico_set_layer("vid");
        int ok = pico_set_video("vid", 0);
        assert(ok == 1);
        pico_set_layer(NULL);
        pico_output_clear();
        pico_output_draw_layer("vid",
            &(Pico_Rel_Rect){ '%', {0.5, 0.5, 1, 1}, PICO_ANCHOR_C, NULL }
        );
        _pico_check("video-01");
    }

    /* pico_set_video: middle frame (25) */
    {
        puts("set_video: frame 25 (bottom-middle)");
        int ok = pico_set_video("vid", 25);
        assert(ok == 1);
        pico_output_clear();
        pico_output_draw_layer("vid",
            &(Pico_Rel_Rect){ '%', {0.5, 0.5, 1, 1}, PICO_ANCHOR_C, NULL }
        );
        _pico_check("video-02");
    }

    /* pico_set_video: last frame (49) */
    {
        puts("set_video: frame 49 (top-right)");
        int ok = pico_set_video("vid", 49);
        assert(ok == 1);
        pico_output_clear();
        pico_output_draw_layer("vid",
            &(Pico_Rel_Rect){ '%', {0.5, 0.5, 1, 1}, PICO_ANCHOR_C, NULL }
        );
        _pico_check("video-03");
    }

    /* pico_set_video: backward seek */
    {
        puts("set_video: backward seek to frame 10");
        int ok = pico_set_video("vid", 10);
        assert(ok == 1);
        pico_output_clear();
        pico_output_draw_layer("vid",
            &(Pico_Rel_Rect){ '%', {0.5, 0.5, 1, 1}, PICO_ANCHOR_C, NULL }
        );
        _pico_check("video-04");
    }

    /* pico_set_video: past EOF */
    {
        puts("set_video: past EOF returns 0");
        int ok = pico_set_video("vid", 100);
        assert(ok == 0);
    }

    /* pico_get_video: check state after sync */
    {
        puts("get_video: state after sync");
        Pico_Video info = pico_get_video("vid", NULL);
        assert(info.done == 1);
    }

    /* pico_output_draw_video: all-in-one */
    {
        puts("draw_video: first call");
        pico_output_clear();
        int ok = pico_output_draw_video("video.y4m",
            &(Pico_Rel_Rect){ '%', {0.5, 0.5, 1, 1}, PICO_ANCHOR_C, NULL }
        );
        assert(ok == 1);
        _pico_check("video-05");
    }

    pico_init(0);
    return 0;
}
